src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs

```diff
@@ -98,7 +98,7 @@ public static string GetUrl(string suffix)
         /// <param name="timeout">An integer indicating the number of milliseconds to wait before timing out while downloading a resource.</param>
         /// <returns>The download results, containing the file path where the resources was (or should have been) downloaded to, and an error message
         /// (or null if there was no error).</returns>
-        public async Task<ResourceDownloadResults> EnsureResource(IHostEnvironment env, IChannel ch, string relativeUrl, string fileName, string dir, int timeout)
```

Comment:
Is there a particular reason to no change the interface from returning a Task to making a synchronous call?

Reply:
change to synchronous call to retry on download, we need to wait for the last result of download (success or fail) then determine whether to return result or continue retry download.
this is not public API we only use this method as internal and not intended to let our user use this

---
In reply to: [375020245](https://github.com/dotnet/machinelearning/pull/4776#discussion_r375020245) [](ancestors = 375020245)

Reply:
Even if you have to wait for the last download to complete and examine the status before deciding whether to continue retrying, you should still use the async/await pattern. If you don't you might block the calling thread.

---
In reply to: [375064673](https://github.com/dotnet/machinelearning/pull/4776#discussion_r375064673) [](ancestors = 375064673,375020245)